armory/scenarios/poison.py

```diff
@@ -518,28 +532,116 @@ def evaluate_current(self):
+        """
+        if self.explanatory_model is None:
+            raise ValueError("No explanatory model")
+
+        class_majority_mask = metrics.get("class_majority_mask")
+        activations = self.explanatory_model.get_activations(
+            self.x_poison[~self.poisoned]
+        )
+        (
+            self.majority_mask_train_unpoisoned,
+            self.majority_ceilings,
+        ) = class_majority_mask(
+            activations,
+            self.y_poison[~self.poisoned],
+        )
+        return self.majority_mask_train_unpoisoned, self.majority_ceilings
+
+    def get_test_majority_mask(self):
+        """
+        get majority ceilings on unpoisoned part of test set
+        """
+        if self.explanatory_model is None:
+            raise ValueError("No explanatory model")
+        if not hasattr(self, "majority_ceilings"):
+            raise ValueError("Must first call 'get_train_majority_mask_and_ceilings'")
+        class_majority_mask = metrics.get("class_majority_mask")
+        activations = self.explanatory_model.get_activations(self.test_x)
+        # use copy of majority ceilings computed from train set
+        self.majority_mask_test_set, _ = class_majority_mask(
+            activations,
+            self.test_y,
+            majority_ceilings=copy.copy(self.majority_ceilings),
+        )
+
+        return self.majority_mask_test_set
+
+    def record_bias_results(self, chi2_spd, class_ids, record_prefix, log_prefix):
+        for class_id in class_ids:
+            chi2, spd = chi2_spd[class_id]
+            self.hub.record(
+                f"{record_prefix}bias_chi^2_p_value_{str(class_id).zfill(2)}", chi2
+            )
+            self.hub.record(f"{record_prefix}bias_spd_{str(class_id).zfill(2)}", spd)
+        log.info(
+            f"{log_prefix}Subclass Bias for Class {str(class_id).zfill(2)}: chi^2 p-value = {chi2:.4f}"
+        )
+        log.info(
+            f"{log_prefix}Subclass Bias for Class {str(class_id).zfill(2)}: SPD = {spd:.4f}"
+        )
+
+    def compute_explanatory(self):
+        log.info("Computing fairness metrics")
+        if self.train_set_class_labels != self.test_set_class_labels:
+            log.warning(
+                "Test set contains a strict subset of train set classes. "
+                "Some metrics for missing classes may not be computed."
+            )
+
+        class_bias = metrics.get("class_bias")
+        # Model bias: compares rate of correct predictions between binary clusters of each class
+
+        self.get_train_majority_mask_and_ceilings()
+        self.get_test_majority_mask()
+        chi2_spd = class_bias(
+            self.test_y,
+            self.majority_mask_test_set,
+            self.test_y == self.test_y_pred_class,
+            self.test_set_class_labels,
+        )
+
+        self.record_bias_results(
+            chi2_spd, self.test_set_class_labels, "model_", "Model "
+        )
+        if self.is_filtering_defense:
```

Comment:
self.is_filtering_defense has not been defined.

Reply:
should be `self.use_filtering_defense`

Reply:
Done.